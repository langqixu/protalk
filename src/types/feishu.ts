// 飞书事件类型
export interface FeishuEvent {
  type: 'url_verification' | 'event_callback' | 'message' | 'command';
  challenge?: string;
  event?: {
    type: string;
    message?: FeishuMessage;
    command?: FeishuCommand;
  };
}

// 飞书消息类型
export interface FeishuMessage {
  message_id: string;
  chat_id: string;
  content: string;
  sender: {
    sender_id: string;
    sender_type: string;
  };
  thread_id?: string;
  create_time?: string;
  card?: any; // 卡片消息数据
}

// 飞书指令类型
export interface FeishuCommand {
  command: string;
  text: string;
  user_id: string;
  chat_id: string;
  message_id: string;
  thread_id?: string;
}

// 飞书配置类型
export interface FeishuConfig {
  // 基础配置
  appId: string;
  appSecret: string;
  webhookUrl: string;
  verificationToken?: string;
  encryptKey?: string;
  
  // 连接配置
  mode: 'webhook' | 'eventsource' | 'hybrid';
  eventSourceUrl?: string;
  connectionTimeout?: number;
  
  // 性能配置
  batchSize?: number;
  retryAttempts?: number;
  retryDelay?: number;
  
  // 队列配置
  queueSize?: number;
  processInterval?: number;
}

// 消息任务类型
export interface MessageTask {
  id: string;
  message: FeishuMessage;
  type: 'new' | 'update' | 'reply' | 'notification';
  timestamp: number;
  retryCount: number;
}

// 队列状态类型
export interface QueueStatus {
  size: number;
  processing: boolean;
  lastProcessed: Date;
  errorCount: number;
  successCount: number;
}

// 连接模式类型
export type ConnectionMode = 'webhook' | 'eventsource' | 'hybrid';

// 模式状态类型
export interface ModeStatus {
  currentMode: ConnectionMode;
  availableModes: ConnectionMode[];
  lastSwitch: Date;
  switchCount: number;
}
