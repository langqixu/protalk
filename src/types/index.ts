// 基于App Store Connect API真实结构的评论数据类型
export interface AppReview {
  // 主键和标识
  reviewId: string;            // App Store Connect API的唯一ID，也是数据库主键
  appId: string;
  
  // App Store Connect API原始字段（保持命名一致）
  rating: number;              // 1-5星评分
  title?: string | null;       // 评论标题（API可选）
  body?: string | null;        // 评论内容
  reviewerNickname: string;    // 评论者昵称（API原始字段名）
  createdDate: Date;           // 评论创建时间（API原始字段名）
  isEdited: boolean;           // 是否编辑过（API原始字段）
  
  // 开发者回复数据（来自customerReviewResponses）
  responseBody?: string | null; // 开发者回复内容
  responseDate?: Date | null;   // 开发者回复时间
  
  // 二级定义字段（业务逻辑分类）
  // dataType 字段已移除，不再区分 review 和 rating_only
  
  // 同步控制字段
  firstSyncAt: Date;           // 首次同步时间
  isPushed: boolean;           // 是否已推送
  pushType?: 'new' | 'historical' | 'updated';
  
  // 扩展字段（预留）
  territoryCode?: string;      // 国家/地区代码
  appVersion?: string;         // 应用版本
  reviewState?: string;        // 评论状态
  
  // 审计字段
  createdAt: Date;
  updatedAt: Date;
}

// Review接口已废弃，统一使用AppReview

// AppFeedback类型已移除，统一使用AppReview

// 评分统计数据结构
export interface AppRatingDailyStats {
  id: string;
  appId: string;
  statDate: string;            // YYYY-MM-DD格式
  
  // 评分分布
  ratingDistribution: {
    1: number;
    2: number;
    3: number;
    4: number;
    5: number;
  };
  
  // 统计指标
  totalRatings: number;        // 当日总评分数
  totalReviews: number;        // 当日评论数
  averageRating: number;       // 当日平均分
  
  // 趋势对比
  ratingChangeFromYesterday: number;  // 平均分环比变化
  ratingCountChange: number;          // 评分数量变化
  reviewCountChange: number;          // 评论数量变化
  
  // 推送状态
  isReportSent: boolean;       // 是否已发送日报
  reportSentAt?: Date;         // 日报发送时间
  
  // 审计字段
  createdAt: Date;
  updatedAt: Date;
}

// 商店配置
export interface StoreConfig {
  type: 'appstore' | 'googleplay';
  appId: string;
  enabled: boolean;
  name: string;
}

// 应用配置
export interface AppConfig {
  stores: StoreConfig[];
  sync: {
    interval?: string;
    batchSize?: number;
    maxRetries?: number;
    retryDelay?: number;
    reviews: {
      interval: string;
      batchSize: number;
      maxRetries: number;
      retryDelay: number;
      enableIncremental: boolean;
      pushNewReviews: boolean;
      pushUpdatedReviews: boolean;
      pushHistoricalReviews: boolean;
      markHistoricalAsPushed: boolean;
    };
    ratings: {
      dailyReportTime: string;
      aggregationPeriod: string;
      pushDailyReport: boolean;
      includeTrends: boolean;
      trendDays: number;
      enabled: boolean;
    };
  };
  api: {
    rateLimit: number;
    timeout: number;
  };
  features?: {
    unifiedDataModel?: boolean;
    smartPushNotifications?: boolean;
    contentChangeDetection?: boolean;
    historicalDataMigration?: boolean;
  };
}

// 环境配置
export interface EnvConfig {
  appStore: {
    issuerId: string;
    keyId: string;
    privateKey: string;
  };
  supabase: {
    url: string;
    anonKey: string;
  };
  feishu: {
    mode?: 'webhook' | 'eventsource';
    // API版本固定为v1，移除配置选项
    enableSignatureVerification?: boolean;
    appId?: string;
    appSecret?: string;
    verificationToken?: string;
    encryptKey?: string;
    batchSize?: number;
    retryAttempts?: number;
    processInterval?: number;
  };
  server: {
    port: number;
    apiKey?: string | undefined;
  };
}

// 统一的反馈抓取器接口
export interface IReviewFetcher {
  syncReviews(appId: string): Promise<AppReview[]>;  // 统一使用AppReview
  replyToReview(reviewId: string, responseBody: string): Promise<{ success: boolean; responseDate: Date }>;
}

// 数据库管理器接口（基于新的AppReview结构）
export interface IDatabaseManager {
  // 基础方法（基于新的AppReview结构）
  upsertAppReviews(reviews: AppReview[]): Promise<void>;
  getExistingReviewIds(appId: string): Promise<Set<string>>;
  getAppReviewsByIds(reviewIds: string[]): Promise<Map<string, AppReview>>;
  getLastSyncTime(appId: string): Promise<Date | null>;
  updateSyncTime(appId: string): Promise<void>;
  
  // 回复相关方法
  updateReply(reviewId: string, responseBody: string, responseDate: Date): Promise<void>;
  hasReply(reviewId: string): Promise<boolean>;
  
  // 数据类型查询方法 - 已移除，不再区分 review 和 rating_only
  // getReviewsByDataType 和 getReviewCountByType 方法已移除
  
  // 统一方法（移除向下兼容）
  
  // 评分统计相关方法
  saveDailyRatingStats?(stats: AppRatingDailyStats): Promise<void>;
  getDailyRatingStats?(appId: string, date: string): Promise<AppRatingDailyStats | null>;
  getRecentRatingTrends?(appId: string, days: number): Promise<AppRatingDailyStats[]>;
}

// 统一的推送器接口
export interface IPusher {
  // 统一使用AppReview
  pushReviewUpdate(review: AppReview, type: 'new' | 'update' | 'reply'): Promise<void>;
  pushBatchUpdates(reviews: AppReview[], type: 'new' | 'update' | 'reply'): Promise<void>;
  
  // 扩展功能
  pushDailyRatingReport?(appId: string, stats: AppRatingDailyStats): Promise<void>;
}

// 统一的数据处理器结果
export interface ProcessedReviews {
  new: AppReview[];
  updated: AppReview[];
}

// 同步结果类型
export interface SyncResult {
  total: number;
  new: number;
  updated: number;
  errors: string[];
}

// 评分日报推送结果
export interface DailyReportResult {
  success: boolean;
  reportSentAt: Date;
  error?: string;
}

// API 响应类型
export interface ApiResponse<T = any> {
  success: boolean;
  data?: T;
  error?: string;
}

// 回复请求类型
export interface ReplyRequest {
  review_id: string;
  response_body: string;
}

// 回复响应类型
export interface ReplyResponse {
  success: boolean;
  responseDate: Date;
}
