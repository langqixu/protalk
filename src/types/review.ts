/**
 * @file review.ts
 * @description Defines the data transfer objects (DTOs) and state enums for app reviews.
 * This file serves as the single source of truth for review-related data structures.
 */

/**
 * Defines the possible states of an interactive review card.
 */
export enum CardState {
  /**
   * Initial state for a new, unreplied review.
   * Displays a "Reply" button.
   */
  NO_REPLY = 'no_reply',

  /**
   * State when the user is actively composing a reply.
   * Displays a text input form with "Submit" and "Cancel" buttons.
   */
  REPLYING = 'replying',

  /**
   * State after a reply has been submitted.
   * Displays the developer's response and an "Edit" button.
   */
  REPLIED = 'replied',

  /**
   * State when the user is editing a previously submitted reply.
   * Displays a pre-filled text input form with "Update" and "Cancel" buttons.
   */
  EDITING_REPLY = 'editing_reply',
}

/**
 * 回复状态枚举
 */
export enum ReplyStatus {
  NONE = 'none',              // 无回复
  PENDING = 'pending',        // 提交中
  SUBMITTED = 'submitted',    // 已提交，等待发布
  PUBLISHED = 'published',    // 已发布
  FAILED = 'failed',          // 提交失败
  UPDATING = 'updating',      // 更新中
  DELETED = 'deleted'         // 已删除
}

/**
 * 应用商店特定数据
 */
export interface StoreSpecificData {
  // App Store Connect特定字段
  appstore?: {
    reviewResponseId?: string;    // 开发者回复的ID
    customerReviewId?: string;    // 原评论在ASC中的ID
    bundleId?: string;           // 应用Bundle ID
  };
  
  // Google Play特定字段（预留）
  googleplay?: {
    packageName?: string;        // 应用包名
    developerReplyId?: string;   // 开发者回复ID
    clusterId?: string;          // 评论聚类ID
  };
}

/**
 * 回复状态追踪
 */
export interface ReplyStatusInfo {
  status: ReplyStatus;           // 当前状态
  lastAttempt?: string;          // 最后尝试时间 (ISO 8601)
  errorMessage?: string;         // 错误信息
  retryCount?: number;           // 重试次数
  submittedAt?: string;          // 提交时间 (ISO 8601)
  publishedAt?: string;          // 发布时间 (ISO 8601)
}

/**
 * Data Transfer Object for an App Store Review.
 * This unified interface is used across the application to ensure consistency.
 */
export interface ReviewDTO {
  // --- Core Identifiers ---
  id: string;
  appId: string;
  appName: string;

  // --- Review Content ---
  rating: number; // 1 to 5
  title: string;
  body: string;
  author: string;
  
  // --- Metadata ---
  createdAt: string; // ISO 8601 date string
  version: string;
  countryCode: string; // e.g., "CN", "US"

  // --- Developer Response ---
  developerResponse?: {
    body: string;
    lastModified: string; // ISO 8601 date string
  };
  
  // --- Store Integration (新增) ---
  storeType?: 'appstore' | 'googleplay'; // 应用商店类型
  storeSpecificData?: StoreSpecificData;  // 商店特定数据
  replyStatus?: ReplyStatusInfo;          // 回复状态信息
  
  // --- Card-specific data (optional) ---
  messageId?: string;
}
