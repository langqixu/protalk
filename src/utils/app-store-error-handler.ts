/**
 * App Store Connect API错误处理工具
 * 提供详细的错误信息和用户友好的错误消息
 */

import logger from './logger';

export interface AppStoreError {
  code: string;
  detail: string;
  status: string;
  title: string;
  source?: {
    pointer?: string;
    parameter?: string;
  };
}

export interface AppStoreErrorResponse {
  errors: AppStoreError[];
}

export class AppStoreErrorHandler {
  /**
   * 处理App Store Connect API错误
   */
  static handleError(error: any): {
    userMessage: string;
    technicalMessage: string;
    errorCode: string;
    retryable: boolean;
  } {
    // 默认错误信息
    let userMessage = '提交回复时发生错误，请稍后重试';
    let technicalMessage = error.message || '未知错误';
    let errorCode = 'UNKNOWN_ERROR';
    let retryable = true;

    if (error.response?.data?.errors && error.response.data.errors.length > 0) {
      const apiErrors: AppStoreError[] = error.response.data.errors;
      const firstError = apiErrors[0];
      
      if (firstError) {
        errorCode = firstError.code;
        technicalMessage = firstError.detail;
        
        // 根据具体错误代码提供用户友好的消息
        switch (firstError.code) {
        case 'FORBIDDEN':
          userMessage = '没有权限回复此评论，请检查App Store Connect账户权限';
          retryable = false;
          break;
          
        case 'NOT_FOUND':
          userMessage = '找不到指定的评论，可能已被删除';
          retryable = false;
          break;
          
        case 'DUPLICATE_RESPONSE':
          userMessage = '此评论已有回复，无法重复回复';
          retryable = false;
          break;
          
        case 'INVALID_RESPONSE_BODY':
          userMessage = '回复内容不符合要求，请检查内容长度和格式';
          retryable = false;
          break;
          
        case 'RATE_LIMITED':
          userMessage = '请求过于频繁，请稍后重试';
          retryable = true;
          break;
          
        case 'UNAUTHORIZED':
          userMessage = 'App Store Connect认证已过期，请重新配置';
          retryable = true;
          break;
          
        case 'BAD_REQUEST':
          if (technicalMessage.includes('responseBody')) {
            userMessage = '回复内容格式错误，请检查内容';
          } else {
            userMessage = '请求参数错误，请检查回复内容';
          }
          retryable = false;
          break;
          
        case 'INTERNAL_SERVER_ERROR':
          userMessage = 'Apple服务器暂时不可用，请稍后重试';
          retryable = true;
          break;
          
        default:
          userMessage = `提交回复失败：${firstError.title || technicalMessage}`;
          retryable = true;
        }
        
        logger.error('App Store API错误详情', {
          errorCode: firstError.code,
          status: firstError.status,
          title: firstError.title,
          detail: firstError.detail,
          source: firstError.source
        });
      }
    } else if (error.response?.status) {
      // HTTP状态码错误
      const status = error.response.status;
      
      switch (status) {
        case 400:
          userMessage = '请求参数错误，请检查回复内容格式';
          errorCode = 'BAD_REQUEST';
          retryable = false;
          break;
          
        case 401:
          userMessage = 'App Store Connect认证失败，请检查配置';
          errorCode = 'UNAUTHORIZED';
          retryable = true;
          break;
          
        case 403:
          userMessage = '没有权限执行此操作，请检查账户权限';
          errorCode = 'FORBIDDEN';
          retryable = false;
          break;
          
        case 404:
          userMessage = '找不到指定的评论';
          errorCode = 'NOT_FOUND';
          retryable = false;
          break;
          
        case 409:
          userMessage = '此评论已有回复，无法重复回复';
          errorCode = 'CONFLICT';
          retryable = false;
          break;
          
        case 422:
          userMessage = '回复内容不符合要求，请检查内容';
          errorCode = 'UNPROCESSABLE_ENTITY';
          retryable = false;
          break;
          
        case 429:
          userMessage = '请求过于频繁，请稍后重试';
          errorCode = 'RATE_LIMITED';
          retryable = true;
          break;
          
        case 500:
        case 502:
        case 503:
        case 504:
          userMessage = 'Apple服务器暂时不可用，请稍后重试';
          errorCode = 'SERVER_ERROR';
          retryable = true;
          break;
          
        default:
          userMessage = `服务器返回错误 (${status})，请稍后重试`;
          errorCode = `HTTP_${status}`;
          retryable = true;
      }
    } else if (error.code === 'ENOTFOUND' || error.code === 'ECONNREFUSED') {
      userMessage = '网络连接失败，请检查网络设置';
      errorCode = 'NETWORK_ERROR';
      retryable = true;
    } else if (error.code === 'ETIMEDOUT') {
      userMessage = '请求超时，请稍后重试';
      errorCode = 'TIMEOUT';
      retryable = true;
    }

    return {
      userMessage,
      technicalMessage,
      errorCode,
      retryable
    };
  }

  /**
   * 检查错误是否可以重试
   */
  static isRetryable(error: any): boolean {
    const handled = this.handleError(error);
    return handled.retryable;
  }

  /**
   * 获取用户友好的错误消息
   */
  static getUserMessage(error: any): string {
    const handled = this.handleError(error);
    return handled.userMessage;
  }

  /**
   * 记录详细的错误信息
   */
  static logError(error: any, context: { reviewId: string; userId: string; replyContent: string }): void {
    const handled = this.handleError(error);
    
    logger.error('App Store回复API调用失败', {
      ...context,
      errorCode: handled.errorCode,
      userMessage: handled.userMessage,
      technicalMessage: handled.technicalMessage,
      retryable: handled.retryable,
      httpStatus: error.response?.status,
      requestUrl: error.config?.url,
      requestMethod: error.config?.method
    });
  }
}

export default AppStoreErrorHandler;
