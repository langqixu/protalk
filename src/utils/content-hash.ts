import crypto from 'crypto';
import { AppReview } from '../types';

/**
 * 简化的内容变更检测器
 * 统一使用AppReview接口，移除向下兼容代码
 */
export class SimpleChangeDetector {
  /**
   * 为AppReview对象生成内容哈希
   */
  static generateContentHash(feedback: AppReview): string {
    // 包含可能变化的字段
    const contentString = JSON.stringify({
      rating: feedback.rating,
      title: feedback.title,
      body: feedback.body,
      reviewerNickname: feedback.reviewerNickname,
      isEdited: feedback.isEdited,
      responseBody: feedback.responseBody,
      responseDate: feedback.responseDate?.toISOString()
    });
    
    return crypto.createHash('md5').update(contentString).digest('hex');
  }

  /**
   * 检测两个AppReview对象之间的变更
   */
  static detectChanges(
    newFeedback: AppReview, 
    existingFeedback?: AppReview
  ): {
    hasChanged: boolean;
    changes: string[];
    newContentHash: string;
    existingContentHash?: string;
  } {
    const newContentHash = this.generateContentHash(newFeedback);
    
    if (!existingFeedback) {
      return {
        hasChanged: true,
        changes: ['new_feedback'],
        newContentHash,
        existingContentHash: undefined
      };
    }
    
    const existingContentHash = this.generateContentHash(existingFeedback);
    const hasChanged = newContentHash !== existingContentHash;
    
    const changes: string[] = [];
    if (hasChanged) {
      // 检测具体变更字段
      if (newFeedback.rating !== existingFeedback.rating) {
        changes.push('rating');
      }
      if (newFeedback.title !== existingFeedback.title) {
        changes.push('title');
      }
      if (newFeedback.body !== existingFeedback.body) {
        changes.push('body');
      }
      if (newFeedback.reviewerNickname !== existingFeedback.reviewerNickname) {
        changes.push('reviewerNickname');
      }
      if (newFeedback.isEdited !== existingFeedback.isEdited) {
        changes.push('isEdited');
      }
      if (newFeedback.responseBody !== existingFeedback.responseBody) {
        changes.push('responseBody');
      }
      if (newFeedback.responseDate?.getTime() !== existingFeedback.responseDate?.getTime()) {
        changes.push('responseDate');
      }
    }
    
    return {
      hasChanged,
      changes,
      newContentHash,
      existingContentHash
    };
  }

  /**
   * 批量检测变更
   */
  static detectBatchChanges(
    newFeedbacks: AppReview[],
    existingFeedbacks: Map<string, AppReview>
  ): {
    newItems: AppReview[];
    changedItems: AppReview[];
    unchangedItems: AppReview[];
    summary: {
      total: number;
      new: number;
      changed: number;
      unchanged: number;
    };
  } {
    const newItems: AppReview[] = [];
    const changedItems: AppReview[] = [];
    const unchangedItems: AppReview[] = [];

    for (const newFeedback of newFeedbacks) {
      const existing = existingFeedbacks.get(newFeedback.reviewId);
      const result = this.detectChanges(newFeedback, existing);
      
      if (!existing) {
        newItems.push(newFeedback);
      } else if (result.hasChanged) {
        changedItems.push(newFeedback);
      } else {
        unchangedItems.push(newFeedback);
      }
    }

    return {
      newItems,
      changedItems,
      unchangedItems,
      summary: {
        total: newFeedbacks.length,
        new: newItems.length,
        changed: changedItems.length,
        unchanged: unchangedItems.length
      }
    };
  }
}