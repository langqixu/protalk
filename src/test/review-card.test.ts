import { buildReviewCardV2 } from '../utils/feishu-card-v2-builder';
import { CardState, ReviewDTO } from '../types/review';

describe('buildReviewCardV2', () => {
  const mockReview: ReviewDTO = {
    id: 'test_review_123',
    appId: 'test_app_456',
    appName: 'Test App',
    rating: 4,
    title: 'Great App!',
    body: 'This is a test review.',
    author: 'Test User',
    createdAt: '2023-10-27T10:00:00Z',
    version: '1.0.0',
    countryCode: 'US',
    developerResponse: {
      body: 'Thank you for your feedback!',
      lastModified: '2023-10-27T11:00:00Z',
    },
  };

  it('should match snapshot for NO_REPLY state', () => {
    const card = buildReviewCardV2(mockReview, CardState.NO_REPLY);
    expect(card).toMatchSnapshot();
  });

  it('should match snapshot for REPLYING state', () => {
    const card = buildReviewCardV2(mockReview, CardState.REPLYING);
    expect(card).toMatchSnapshot();
  });

  it('should match snapshot for REPLIED state', () => {
    const card = buildReviewCardV2(mockReview, CardState.REPLIED);
    expect(card).toMatchSnapshot();
  });

  it('should match snapshot for EDITING_REPLY state', () => {
    const card = buildReviewCardV2(mockReview, CardState.EDITING_REPLY);
    expect(card).toMatchSnapshot();
  });
});
