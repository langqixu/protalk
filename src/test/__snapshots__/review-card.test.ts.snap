// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildReviewCardV2 should match snapshot for EDITING_REPLY state 1`] = `
{
  "config": {
    "update_multi": true,
    "wide_screen_mode": true,
  },
  "elements": [
    {
      "fields": [
        {
          "is_short": false,
          "text": {
            "content": "⭐⭐⭐⭐☆ (4/5)",
            "tag": "lark_md",
          },
        },
        {
          "is_short": false,
          "text": {
            "content": "**Great App!**

This is a test review.",
            "tag": "lark_md",
          },
        },
        {
          "is_short": false,
          "text": {
            "content": "
",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">👤 **用户:** Test User</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">🌍 **地区:** US</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">📱 **版本:** 1.0.0</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">📅 **时间:** 2023/10/27 18:00:00</font>",
            "tag": "lark_md",
          },
        },
      ],
      "tag": "div",
    },
    {
      "tag": "hr",
    },
    {
      "elements": [
        {
          "default_value": "Thank you for your feedback!",
          "input_type": "multiline_text",
          "name": "reply_content",
          "placeholder": {
            "content": "修改您的回复内容...",
            "tag": "plain_text",
          },
          "required": true,
          "rows": 3,
          "tag": "input",
          "width": "fill",
        },
        {
          "columns": [
            {
              "elements": [
                {
                  "action_type": "form_submit",
                  "name": "update_button",
                  "tag": "button",
                  "text": {
                    "content": "更新",
                    "tag": "plain_text",
                  },
                  "type": "primary",
                  "value": {
                    "action": "replied",
                    "review_id": "test_review_123",
                  },
                  "width": "default",
                },
              ],
              "horizontal_align": "left",
              "tag": "column",
              "vertical_align": "top",
              "vertical_spacing": "8px",
              "width": "auto",
            },
            {
              "elements": [
                {
                  "action_type": "request",
                  "name": "cancel_edit_button",
                  "tag": "button",
                  "text": {
                    "content": "取消",
                    "tag": "plain_text",
                  },
                  "type": "default",
                  "value": {
                    "action": "replied",
                    "review_id": "test_review_123",
                  },
                  "width": "default",
                },
              ],
              "horizontal_align": "left",
              "tag": "column",
              "vertical_align": "top",
              "vertical_spacing": "8px",
              "width": "auto",
            },
          ],
          "horizontal_align": "left",
          "tag": "column_set",
        },
      ],
      "name": "edit_reply_form",
      "tag": "form",
    },
  ],
  "header": {
    "template": "green",
    "title": {
      "content": "✅ Test App - 新评论通知",
      "tag": "plain_text",
    },
  },
}
`;

exports[`buildReviewCardV2 should match snapshot for NO_REPLY state 1`] = `
{
  "config": {
    "update_multi": true,
    "wide_screen_mode": true,
  },
  "elements": [
    {
      "fields": [
        {
          "is_short": false,
          "text": {
            "content": "⭐⭐⭐⭐☆ (4/5)",
            "tag": "lark_md",
          },
        },
        {
          "is_short": false,
          "text": {
            "content": "**Great App!**

This is a test review.",
            "tag": "lark_md",
          },
        },
        {
          "is_short": false,
          "text": {
            "content": "
",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">👤 **用户:** Test User</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">🌍 **地区:** US</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">📱 **版本:** 1.0.0</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">📅 **时间:** 2023/10/27 18:00:00</font>",
            "tag": "lark_md",
          },
        },
      ],
      "tag": "div",
    },
    {
      "tag": "hr",
    },
    {
      "actions": [
        {
          "behaviors": [
            {
              "type": "callback",
              "value": {
                "action": "replying",
                "review_id": "test_review_123",
              },
            },
          ],
          "tag": "button",
          "text": {
            "content": "回复",
            "tag": "plain_text",
          },
          "type": "primary",
        },
      ],
      "tag": "action",
    },
  ],
  "header": {
    "template": "green",
    "title": {
      "content": "✅ Test App - 新评论通知",
      "tag": "plain_text",
    },
  },
}
`;

exports[`buildReviewCardV2 should match snapshot for REPLIED state 1`] = `
{
  "config": {
    "update_multi": true,
    "wide_screen_mode": true,
  },
  "elements": [
    {
      "fields": [
        {
          "is_short": false,
          "text": {
            "content": "⭐⭐⭐⭐☆ (4/5)",
            "tag": "lark_md",
          },
        },
        {
          "is_short": false,
          "text": {
            "content": "**Great App!**

This is a test review.",
            "tag": "lark_md",
          },
        },
        {
          "is_short": false,
          "text": {
            "content": "
",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">👤 **用户:** Test User</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">🌍 **地区:** US</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">📱 **版本:** 1.0.0</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">📅 **时间:** 2023/10/27 18:00:00</font>",
            "tag": "lark_md",
          },
        },
      ],
      "tag": "div",
    },
    {
      "tag": "hr",
    },
    {
      "extra": {
        "behaviors": [
          {
            "type": "callback",
            "value": {
              "action": "editing_reply",
              "review_id": "test_review_123",
            },
          },
        ],
        "tag": "button",
        "text": {
          "content": "编辑回复",
          "tag": "plain_text",
        },
        "type": "default",
      },
      "fields": [
        {
          "is_short": false,
          "text": {
            "content": "**开发者回复:**
Thank you for your feedback!",
            "tag": "lark_md",
          },
        },
      ],
      "tag": "div",
    },
  ],
  "header": {
    "template": "green",
    "title": {
      "content": "✅ Test App - 新评论通知",
      "tag": "plain_text",
    },
  },
}
`;

exports[`buildReviewCardV2 should match snapshot for REPLYING state 1`] = `
{
  "config": {
    "update_multi": true,
    "wide_screen_mode": true,
  },
  "elements": [
    {
      "fields": [
        {
          "is_short": false,
          "text": {
            "content": "⭐⭐⭐⭐☆ (4/5)",
            "tag": "lark_md",
          },
        },
        {
          "is_short": false,
          "text": {
            "content": "**Great App!**

This is a test review.",
            "tag": "lark_md",
          },
        },
        {
          "is_short": false,
          "text": {
            "content": "
",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">👤 **用户:** Test User</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">🌍 **地区:** US</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">📱 **版本:** 1.0.0</font>",
            "tag": "lark_md",
          },
        },
        {
          "is_short": true,
          "text": {
            "content": "<font color="grey">📅 **时间:** 2023/10/27 18:00:00</font>",
            "tag": "lark_md",
          },
        },
      ],
      "tag": "div",
    },
    {
      "tag": "hr",
    },
    {
      "elements": [
        {
          "input_type": "multiline_text",
          "name": "reply_content",
          "placeholder": {
            "content": "输入您对用户的回复内容...",
            "tag": "plain_text",
          },
          "required": true,
          "rows": 3,
          "tag": "input",
          "width": "fill",
        },
        {
          "columns": [
            {
              "elements": [
                {
                  "action_type": "form_submit",
                  "name": "submit_button",
                  "tag": "button",
                  "text": {
                    "content": "提交",
                    "tag": "plain_text",
                  },
                  "type": "primary",
                  "value": {
                    "action": "replied",
                    "review_id": "test_review_123",
                  },
                  "width": "default",
                },
              ],
              "horizontal_align": "left",
              "tag": "column",
              "vertical_align": "top",
              "vertical_spacing": "8px",
              "width": "auto",
            },
            {
              "elements": [
                {
                  "action_type": "request",
                  "name": "cancel_button",
                  "tag": "button",
                  "text": {
                    "content": "取消",
                    "tag": "plain_text",
                  },
                  "type": "default",
                  "value": {
                    "action": "no_reply",
                    "review_id": "test_review_123",
                  },
                  "width": "default",
                },
              ],
              "horizontal_align": "left",
              "tag": "column",
              "vertical_align": "top",
              "vertical_spacing": "8px",
              "width": "auto",
            },
          ],
          "horizontal_align": "left",
          "tag": "column_set",
        },
      ],
      "name": "reply_form",
      "tag": "form",
    },
  ],
  "header": {
    "template": "green",
    "title": {
      "content": "✅ Test App - 新评论通知",
      "tag": "plain_text",
    },
  },
}
`;
